sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,387 @@
+option `-fsycl-link`, where `-fsycl-link=image` creates AOCX archives and
+`-fsycl-link=early` generates AOCR archives.  The files generated by these
+options are handled in a specific manner when encountered.
+
+Any archive with an AOCR type device binary will have the AOCR binary
+extracted and passed to `aoc` to produce an AOCX final image.  This final
+image is wrapped and added to the final binary during the host link.  The use
+of `-fsycl-link=image` with an AOCR binary will create an AOCX based archive
+instead of completing the host link.  Any archive with an AOCX type device
+binary skips the `aoc` step and is wrapped and added to the final binary during
+the host link.  Archives with any AOCO device binaries are extracted and passed
+through to `aoc -library-list=<listfile>`
+
+As the `clang-linker-wrapper` is responsible for understanding the archives
+that are added on the command line, it will need to know when to look for
+these unique device binaries based on the expected compilation targets.  The
+behavior of creating the AOCX/AOCR type archive will be triggered via an
+additional command line option specified by the driver when `-fsycl-link`
+options are used.  The `--fpga-link=<type>` option will tell the wrapper when
+these handlings need to occur.
+
+When using the `-fintelfpga` option to enable AOT for FPGA, there are
+additional expectations during the compilation.  Use of the option will enable
+debug generation and also generate dependency information.  The dependency
+generation should be packaged along with the device binary for use during
+the link phase.  It is expected that the full fat object, containing host,
+device and dependency file is generated before being passed to the link phase.
+The dependency information is only used when compiling for hardware.
+
+The `clang-linker-wrapper` tool will be responsible to determine which FPGA
+tool is being used during the AOT device compilation phase.  The use of
+`-simulation` or `-hardware` as passed in by `--fpga-tool-arg` signifies
+which tool is used.
+
+#### spir64_x86_64 support
+
+Compilation behaviors involving AOT for CPU involve an additional call to
+`opencl-aot`.  This call occurs after the post-link step performed by
+`sycl-post-link`.  Additional options passed by the user via the
+`-Xsycl-target-backend=spir64_x86_64 <opts>` command will be processed by a new
+option to the wrapper, `--cpu-tool-arg=<arg>`
+
+### Wrapping of device image
+
+Once the device binary is pulled out of the fat binary, the binary must be
+wrapped and provided the needed entry points to be used during execution.  This
+is performed by the during the link phase and controlled by the
+`clang-linker-wrapper`.
+
+It is expected that the wrap information that is generated to be wrapped
+around the device binary will match current wrapping information that is used
+for the exiting offload model.  The wrapping in the old model is using the
+`clang-offload-wrapper` tool.
```

Comment:
In order to implement the AOT part of "optional kernel features", we need to somehow annotate each device binary with an identifier for its GPU target.  For example, a device binary created for `--gen-tool-arg=pvc` needs to be annotated as a "PVC" binary.  My thought was to use the GMDID for this annotation.

I think we currently annotate AOT binaries with `PI_DEVICE_BINARY_TYPE_NATIVE`.  Is this part of the "wrapping" you describe here?  Is there anything about this design that will make it difficult to change this to the GMDID?

Reply:
We would have to provide the GMDID mapping in the driver (and continue to update as new ID values come along) so when a user wants `intel_gpu_pvc` the GMDID will be used and embedded into corresponding information tagged with the device binary within the fat object.  When the `clang-linker-wrapper` is extracting, it will know to use the information as tagged accordingly to set the proper device with the `ocloc` call.

Reply:
Rather than embedding the list of GMDID's in the driver, I think we can include the GMDID for each target in the device config file that we are designing (#9371).  Once that happens, the driver will use the config file information to determine the set of legal targets for `-fsycl-targets` and also get the associated GMDID from there.